File: tests/src/org.json4s/json4s-core_2.13/3.7.0-M13/src/test/scala/org_json4s/json4s_core_2_13/ReflectorTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_json4s.json4s_core_2_13

import org.assertj.core.api.Assertions.assertThat
import org.json4s.reflect.ClassDescriptor
import org.json4s.reflect.ConstructorParamDescriptor
import org.json4s.reflect.Reflector
import org.junit.jupiter.api.Test

class ReflectorTest {
  @Test
  def invokesDefaultValueAccessorFromDescribedConstructor(): Unit = {
    Reflector.clearCaches()

    val descriptor: ClassDescriptor = Reflector.describe[ReflectorDefaultValueFixture].asInstanceOf[ClassDescriptor]
    val defaultedParam: ConstructorParamDescriptor = descriptor.mostComprehensive.find(_.name == "label").get

    assertThat(defaultedParam.hasDefault).isTrue()
    assertThat(defaultedParam.defaultValue.isDefined).isTrue()
    assertThat(defaultedParam.defaultValue.get.apply()).isEqualTo("from-default")
  }
}

case class ReflectorDefaultValueFixture(label: String = "from-default", count: Int = 7)

File: tests/src/plexus/plexus-utils/1.0-alpha-2/src/test/java/plexus/plexus_utils/ReflectorTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package plexus.plexus_utils;

import java.lang.reflect.Method;

import org.codehaus.plexus.util.reflection.Reflector;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class ReflectorTest {
    @Test
    void createsObjectsUsingMatchingPublicConstructors() throws Exception {
        Reflector reflector = new Reflector();

        Object instance = reflector.newInstance(ReflectorFixture.class, new Object[] {"created"});

        assertThat(instance).isInstanceOf(ReflectorFixture.class);
        assertThat(((ReflectorFixture) instance).value).isEqualTo("created");
    }

    @Test
    void readsPublicStaticAndInstanceFields() throws Exception {
        Reflector reflector = new Reflector();
        ReflectorFixture fixture = new ReflectorFixture("field-value");

        assertThat(reflector.getStaticField(ReflectorFixture.class, "STATIC_VALUE")).isEqualTo("static-value");
        assertThat(reflector.getField(fixture, "value")).isEqualTo("field-value");
    }

    @Test
    void invokesPublicInstanceStaticAndSingletonMethods() throws Exception {
        Reflector reflector = new Reflector();
        ReflectorFixture fixture = new ReflectorFixture("call");

        assertThat(reflector.invoke(fixture, "append", new Object[] {"-suffix"})).isEqualTo("call-suffix");
        assertThat(reflector.invokeStatic(ReflectorFixture.class, "join", new Object[] {"left", "right"}))
                .isEqualTo("left:right");

        Object singleton = reflector.getSingleton(ReflectorFixture.class, new Object[] {"singleton"});

        assertThat(singleton).isInstanceOf(ReflectorFixture.class);
        assertThat(((ReflectorFixture) singleton).value).isEqualTo("singleton");
    }

    @Test
    void locatesPublicNoArgumentAccessor() throws Exception {
        Reflector reflector = new Reflector();

        Method method = reflector.getMethod(ReflectorFixture.class, "getValue", new Class[0]);

        assertThat(method.getName()).isEqualTo("getValue");
        assertThat(method.getParameterTypes()).isEmpty();
    }

    public static class ReflectorFixture {
        public static final String STATIC_VALUE = "static-value";

        public String value;

        public ReflectorFixture() {
            this("default");
        }

        public ReflectorFixture(String value) {
            this.value = value;
        }

        public static ReflectorFixture getInstance(String value) {
            return new ReflectorFixture(value);
        }

        public static String join(String left, String right) {
            return left + ":" + right;
        }

        public String append(String suffix) {
            return value + suffix;
        }

        public String getValue() {
            return value;
        }
    }
}
